broker-daemon/state-machines/interchain.js

```diff
@@ -0,0 +1,310 @@
+ * @param   {Error}  error    - Error that caused the cancel
+ * @returns {void}
+ */
+async function cancelSwap (engine, hash, error) {
+  logger.error('Permanent Error encountered while translating swap, ' +
+    'cancelling upstream invoice', { error: error.message, hash })
+
+  return engine.cancelSwap(hash)
+}
+
+/**
+ * Settle an upstream payment for a swap.
+ *
+ * @private
+ *
+ * @param   {Engine} engine
+ * @param   {string} hash     - Base64 string of the swap hash
+ * @param   {string} preimage - Base64 string of the swap preimage
+ * @returns {void}
+ */
+async function settleSwap (engine, hash, preimage) {
+  logger.debug(`Settling upstream payment for ${hash}`)
+  await engine.settleSwap(preimage)
+  logger.debug(`Successfully settled upstream payment for ${hash}`)
+}
+
+/**
+ * Retry forwarding a swap across chains.
+ *
+ * @private
+ *
+ * @param {string}  hash                  - Base64 string of the hash for the
+ *                                          swap
+ * @param {Payment} inboundPayment        - Expected inbound payment
+ * @param {Payment} outboundPayment       - Outbound payment we will make to
+ *                                          retrieve the preimage.
+ * @param {Error}   error                 - Error that caused the retry
+ * @returns {string}                        Base64 encoded preimage for the swap
+ */
+async function retryForward (hash, inboundPayment, outboundPayment, error) {
+  logger.error('Temporary Error encountered while translating swap',
+    { error: error.stack, hash })
+
+  logger.debug(`Delaying swap retry for ${hash} for ${RETRY_DELAY}ms`)
+  await delay(RETRY_DELAY)
+  logger.debug(`Retrying swap translation for ${hash}`)
+
+  return forwardSwap(hash, inboundPayment, outboundPayment)
+}
+
+/**
+ * Forward a swap cross-chain by retrieving the preimage from the
+ * downstream chain (by paying an invoice to its hash) and returning it
+ * to the upstream chain, cancelling the upstream when an unrecoverable
+ * error is encountered. When any other error is encountered, it retries
+ * to protect us against being in a non-atomic state.
+ *
+ * @public
+ *
+ * @param {string}  hash                  - Base64 string of the hash for the
+ *                                          swap
+ * @param {Payment} inboundPayment        - Expected inbound payment
+ * @param {Payment} outboundPayment       - Outbound payment we will make to
+ *                                          retrieve the preimage.
+ * @returns {string}                        Base64 encoded preimage for the swap
+ * @throws {Error} If a permanent error is encountered and the swap is cancelled
+ */
+async function forwardSwap (hash, inboundPayment, outboundPayment) {
+  try {
+    const paymentPreimage = await translateIdempotent(hash, inboundPayment, outboundPayment)
+    logger.debug(`Successfully retrieved preimage for swap ${hash}`)
+
+    // settle the upstream payment prior to returning the preimage
```

Comment:
[nit] I think the code is probably clear enough to go without a comment here, and same for the `cancelSwap` comment below. (I like the philosophy from "Clean Code": http://www.kyleblaney.com/software-blog/2012/6/29/comments-are-a-failure.html)